Snowflake.Data/Core/Session/SessionPool.cs

```diff
@@ -95,68 +103,170 @@ internal SFSession GetSession(string connStr, SecureString password)
+                    }
+                    s_logger.Debug("SessionPool::GetIdleSession - no one was waiting for session, but could not find any idle session available");
+                    if (IsAllowedToCreateNewSession())
+                    {
+                        // there is no need to wait for a session since we can create a new one
+                        return new SessionOrCreationToken(_sessionCreationTokenCounter.NewToken());
+                    }
+                }
+            }
+            return new SessionOrCreationToken(WaitForSession(connStr));
+        }
+
+        private bool IsAllowedToCreateNewSession()
+        {
+            if (!_waitingForSessionToReuseQueue.IsWaitingEnabled())
+            {
+                s_logger.Debug($"SessionPool - new session creation granted");
+                return true;
+            }
+            var currentSize = GetCurrentPoolSize();
+            if (currentSize < _maxPoolSize)
+            {
+                s_logger.Debug($"SessionPool - new session creation granted because current size is {currentSize} out of {_maxPoolSize}");
+                return true;
+            }
+            s_logger.Debug($"SessionPool - could not grant new session creation because current size is {currentSize} out of {_maxPoolSize}");
+            return false;
+        }
+        
+        private SFSession WaitForSession(string connStr)
+        {
+            var timeout = _waitingForSessionToReuseQueue.GetWaitingTimeoutMillis();
+            s_logger.Info($"SessionPool::WaitForSession for {timeout} millis timeout");
+            _sessionPoolEventHandler.OnWaitingForSessionStarted(this);
+            var beforeWaitingTime = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
+            long nowTime = beforeWaitingTime;
+            while (nowTime < beforeWaitingTime + timeout) // we loop to handle the case if someone overtook us after being woken or session which we were promised has just expired
+            {
+                var timeoutLeft = beforeWaitingTime + timeout - nowTime;
+                _sessionPoolEventHandler.OnWaitingForSessionStarted(this, timeoutLeft);
+                var successful = _waitingForSessionToReuseQueue.Wait((int) timeoutLeft, CancellationToken.None);
+                if (successful)
+                {
+                    s_logger.Debug($"SessionPool::WaitForSession - woken with a session granted");
+                    _sessionPoolEventHandler.OnWaitingForSessionSuccessful(this);
+                    lock (_sessionPoolLock)
+                    {
+                        var session = ExtractIdleSession(connStr);
+                        if (session != null)
                         {
-                            s_logger.Debug($"reuse pooled session with sid {session.sessionId}");
-                            _busySessionsCounter.Increase();
+                            s_logger.Debug($"SessionPool::WaitForSession - a session was extracted from idle sessions");
```

Comment:
extracted - can we change it here as well?

Reply:
done